File: apps/ipad-shell/files/file-token-store.swift
import Foundation

let maximumScoreFileBytes = 64 * 1024 * 1024

struct FileTokenEntry: Sendable {
    let url: URL
    let fileName: String
    let sizeBytes: Int
}

actor FileTokenStore {
    private struct StoredEntry {
        let value: FileTokenEntry
        let expiresAt: Date
    }

    private var entries: [String: StoredEntry] = [:]
    private let maxBytes: Int
    private let ttl: TimeInterval
    private let now: @Sendable () -> Date

    init(
        maxBytes: Int = maximumScoreFileBytes,
        ttl: TimeInterval = 60,
        now: @escaping @Sendable () -> Date = Date.init
    ) {
        self.maxBytes = maxBytes
        self.ttl = ttl
        self.now = now
    }

    var outstandingCount: Int { entries.count }

    func issue(url: URL, fileName: String, sizeBytes: Int) throws -> String {
        try issueBatch([
            FileTokenEntry(url: url, fileName: fileName, sizeBytes: sizeBytes)
        ])[0]
    }

    func issueBatch(_ values: [FileTokenEntry]) throws -> [String] {
        for value in values {
            guard value.sizeBytes >= 0, value.sizeBytes <= maxBytes else {
                throw fileTokenError("FILE_TOO_LARGE")
            }
            guard value.url.isFileURL, !value.fileName.isEmpty else {
                throw fileTokenError("FILE_SELECTION_INVALID")
            }
        }
        let expiresAt = now().addingTimeInterval(ttl * Double(max(1, values.count)))
        return values.map { value in
            let token = UUID().uuidString.lowercased()
            entries[token] = StoredEntry(value: value, expiresAt: expiresAt)
            return token
        }
    }

    func consume(_ token: String) throws -> FileTokenEntry {
        guard let stored = entries.removeValue(forKey: token) else {
            throw fileTokenError("FILE_TOKEN_INVALID")
        }
        guard stored.expiresAt > now() else {
            throw fileTokenError("FILE_TOKEN_EXPIRED")
        }
        return stored.value
    }

    func discard(_ token: String) {
        entries.removeValue(forKey: token)
    }

    func clear() {
        entries.removeAll()
    }
}

private func fileTokenError(_ code: String) -> NSError {
    NSError(
        domain: "com.hikerpig.zupulse.file-token",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: code]
    )
}

File: apps/desktop-shell/src/main/file-access/file-token-store.ts
import { randomUUID } from "node:crypto";

export type FileTokenEntry = {
  path: string;
  fileName: string;
  sizeBytes: number;
  identity?: { dev: number; ino: number; mtimeMs: number };
  expiresAt: number;
};

export class FileTokenStore {
  private readonly entries = new Map<string, FileTokenEntry>();

  constructor(
    private readonly options: { now: () => number; ttlMs: number } = {
      now: () => Date.now(),
      ttlMs: 60_000,
    },
  ) {}

  issue(
    path: string,
    metadata: { fileName: string; sizeBytes: number; identity?: { dev: number; ino: number; mtimeMs: number } },
    ttlMs?: number,
  ): string {
    const token = randomUUID();
    this.entries.set(token, {
      path,
      ...metadata,
      expiresAt: this.options.now() + (ttlMs ?? this.options.ttlMs),
    });
    return token;
  }

  peek(token: string): FileTokenEntry {
    const entry = this.entries.get(token);
    if (!entry || entry.expiresAt < this.options.now()) {
      throw new Error("FILE_TOKEN_INVALID");
    }
    return entry;
  }

  consume(token: string): FileTokenEntry {
    const entry = this.entries.get(token);
    this.entries.delete(token);
    if (!entry || entry.expiresAt < this.options.now()) {
      throw new Error("FILE_TOKEN_INVALID");
    }
    return entry;
  }

  clear(): void {
    this.entries.clear();
  }
}
